Simplify defeat backend

Hdt80bro opened the pull request:
This PR refactors the AIBrain's defeat code. The main change is moving the unit destruction logic to be all in one spot, in `SimUtils.lua`, no matter how the brain was defeated.

In the process, in was discovered that the armies are handled differently in a recall (it transferred only built units before the 10 seconds wait - it now also transfers unbuilt units, and after the 10 grace period like in a regular defeat), so that has now been standardized. `TransferUnitsToKiller` was also using the incorrect callback in one particular case.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Defeat/recall/disconnect flows now transfer units consistently and behave more reliably at end of game.

* **New Features**
  * Utilities for safer unit transfers, timed waits, and targeted handling of unsafe/abandoned commanders.
  * Kennel pause/upgrade and unit-upgrade behaviors for transferred units.
  * Lobby option types refined for clearer share/disconnect choices.
  * End-of-game delay reduced to 1 second (configurable constant).

* **Refactor**
  * Centralized defeat/recall state and consolidated army/unit management paths.

* **Documentation**
  * Added changelog entry documenting the fix.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Hdt80bro committed:
Refactor defeat logic

Hdt80bro committed:
Cleanup

Hdt80bro committed:
Refactor some more

Hdt80bro committed:
Let recall transfer unfinished units

lL1l1 commented on lua/SimUtils.lua:
```diff
@@ -1052,25 +1075,46 @@ function KillArmy(self, shareOption)
----@param shareOption 'FullShare' | 'ShareUntilDeath' | 'PartialShare' | 'TransferToKiller' | 'Defectors' | 'CivilianDeserter'
----@param shareTime number Game time in ticks
+---@param shareOption ShareOption
+---@param shareTime integer Game time in ticks
```
```suggestion
---@param shareTime integer # Game time in ticks
```

Hdt80bro replied:
As I understand, the majority of the repo doesn't use this optional syntax

lL1l1 replied:
It would be nice to clean up the comment style for annotations (using a # or not) in a separate pr then, though that would be low priority.

lL1l1 commented on lua/SimUtils.lua:
```diff
@@ -1163,19 +1208,85 @@ function KillArmyOnACUDeath(self, shareOption)
+        -- non-assassination modes can have armies abandon without commanders
+        if shareAcuOption == 'Recall' and not table.empty(safeCommanders) then
+            -- KillArmy waits 10 seconds before acting, while FakeTeleport waits 3 seconds, so the ACU shouldn't explode.
+            FakeTeleportUnits(safeCommanders, true)
```
No longer forking the teleport as a thread means it always recalls the ACU before possibly killing it, so the comment isn't needed.

It also means that the grace period for an abandoned army is 13 seconds instead of 10. In my opinion that is reasonable.

Hdt80bro replied:
Ah, thanks, I didn't notice, but that is reasonable indeed - it mirrors the recall behavior where the defeat doesn't happen until the recall finishes (which uses the same function), except that the defeat state is set first.

lL1l1 reviewed (commented):
Looks mostly good to me. Needs a snippet and after you unset the draft status I can approve it.

Hdt80bro committed:
Merge branch 'develop' into simplify-defeat-backend

Hdt80bro committed:
Process feedback and more

Hdt80bro committed:
Do another pass

Hdt80bro committed:
Appease the rabbit

lL1l1 commented on lua/SimUtils.lua:
```diff
@@ -740,63 +740,67 @@ end
----@param owner number
--- categoriesToKill is an optional input (it defaults to all categories)
+local defaultTransferCategory = categories.ALLUNITS - categories.WALL - categories.COMMAND
+-- only units in this category will be shared during a partial share
```
"will be shared under partial share rules" is better

lL1l1 commented on lua/SimUtils.lua:
```diff
@@ -740,63 +740,67 @@ end
 --- Given that `deadArmy` just died, redistributes their unit cap based on the scenario options
 ---@param deadArmy integer
 function UpdateUnitCap(deadArmy)
     -- If we are asked to share out unit cap for the defeated army, do the following...
```
unnecessary comment

lL1l1 commented on lua/SimUtils.lua:
```diff
@@ -740,63 +740,67 @@ end
-                caps[aliveCount] = cap
-            end
+        if not ArmyIsCivilian(index) and not brain:IsDefeated() and
+            (modeAll or IsAlly(deadArmy, index))
```
"shareToAll" would be a better variable name

lL1l1 commented on lua/SimUtils.lua:
```diff
@@ -1023,7 +1026,7 @@ local function TransferUnitsToKiller(self)
 
         if killerIndex then
-            TransferUnitsToBrain(self, { ArmyBrains[killerIndex] }, true, nil, "TransferToKiller")
+            TransferUnitsToBrain(self, { GetArmyBrain(killerIndex) }, true, nil, "TransferToKiller")
```
Both of these work; is global table access or an engine function call better?

lL1l1 replied:
Ran a simple benchmark, even a global is 52x faster.

lL1l1 commented on lua/SimUtils.lua:
```diff
@@ -1086,21 +1088,61 @@ function KillRecalledArmy(self, shareOption)
+--- (which could have been negative or fractional) if it was reached after
+--- checking the units. If all units are dead upon calling, returns `0`.
+---@param units Entity[]
+---@param timeout? integer in ticks
```
`timeoutTicks` would be better

lL1l1 commented on lua/SimUtils.lua:
```diff
@@ -1086,21 +1088,61 @@ function KillRecalledArmy(self, shareOption)
+---@param units Entity[]
+---@param timeout? integer in ticks
+---@return integer elapsed
+function ThreadSuspendUntilUnitsDead(units, timeout)
```
`WaitUntilUnitsDeadOrTimeout` would be better because "thread suspension" is closer to what's going on in shield.lua.

lL1l1 commented on lua/aibrain.lua:
The variable is no longer used and can be removed (it is quite new).

lL1l1 commented on lua/aibrain.lua:
```diff
@@ -8,14 +8,10 @@
 local DisableAI = import("/lua/simutils.lua").DisableAI
-local TransferUnitsToBrain = import("/lua/simutils.lua").TransferUnitsToBrain
-local TransferUnitsToHighestBrain = import("/lua/simutils.lua").TransferUnitsToHighestBrain
 local UpdateUnitCap = import("/lua/simutils.lua").UpdateUnitCap
```
Simutils gets imported 5 times so maybe it should be in a variable.

lL1l1 committed:
Fix "undefined global" warnings

lL1l1 committed:
Fix undefined field warning

lL1l1 committed:
Add timestamps to victory condition debug logging

lL1l1 reviewed (commented):
Some testing using the launch script and 4 players (fullshare, assassination):
- Defeat through recalling seems fine, although I know that in the live game that it kills units before the sim stops.
- Defeat through ctrl k seems fine
- Defeat through abandonment kills units before the sim stops. 

There is a ~109 tick difference between `AbandonedByPlayer` being called and the game ending, which varies due to the ACU kill and victory condition running on different threads. The difference in recall between here and live game is probably due to the additional 3 seconds to the grace period.

The ideal solution would be to make recall and abandon work sequentially with the victory condition thread, but for now the grace period can be increased.

lL1l1 committed:
Increase grace period from 10 to 12 seconds

lL1l1 committed:
debug logging

lL1l1 committed:
Revert "debug logging"

This reverts commit b457ca9ef6c73e08d524469eba7387d5cee034d1.

lL1l1 committed:
Revert "Add timestamps to victory condition debug logging"

This reverts commit 7e198358f28129ed7d95e56bf8bf5a178aebc260.

lL1l1 committed:
improve UpdateUnitCap variable names

lL1l1 committed:
Remove unnecessary comment in UpdateUnitCap

lL1l1 committed:
Update partialShareCategory comment

lL1l1 committed:
Remove unused CommanderSafeTime upvalue

lL1l1 committed:
change name of ThreadSuspendUntilUnitsDead

lL1l1 committed:
de-duplicate SimUtils import in aibrain

lL1l1 committed:
change `GetArmyBrain()` to `ArmyBrains[]`

Hdt80bro commented:
Your changes look good to me, here's my recap so we're on the same page:
* Your local variables name are definitely better than the ones I was able to come up with.
* The function name `ThreadSuspendUntilUnitsDead` was intended to mirror the other thread functions like `ThreadSuspend` and `ThreadResume` but reading it back, it does make more sense to mirror the higher level functions like `WaitSeconds` (which it itself does call) instead. I'm torn on whether the `OrTimeout` part is necessary in `WaitUntilUnitsDeadOrTimeout` as it's an optional parameter that most functions would have; I'll defer to your judgment here as you have better names :)
* Comment updates are good; the one you removed was one I copied. There were several like that I deleted, but missed that one.
* I changed `ArmyBrains[]` to `GetArmyBrains()` purely for consistency knowing it was slower, since it looked like the previous author preferred it and it wasn't being called in performance critical paths; changing all of them back is fine.
* Compacting the imports doesn't affect performance since it's called literally once, but it is more readable, so I like that.
* We really should establish commenting best practices. I'll bring it up in Discord, though it's unlikely anyone will have strong opinions...

At a glance, increasing the grace period to 12 seconds seems curious to me because even 10 seconds already seems too long to me for players to have to wait for. If these changes really started blowing the units too soon, I think we should decrease the victory monitor's wait interval instead.

lL1l1 commented:
> I'm torn on whether the `OrTimeout` part is necessary in `WaitUntilUnitsDeadOrTimeout` as it's an optional parameter that most functions would have; I'll defer to your judgment here as you have better names :)

I think it's nice to have in the name because as you said, most functions would have it. It also makes more clear what the options are for using the function.

> At a glance, increasing the grace period to 12 seconds seems curious to me because even 10 seconds already seems too long to me for players to have to wait for. If these changes really started blowing the units too soon, I think we should decrease the victory monitor's wait interval instead.

There's a 5 second delay that checks for draws, and the wait interval is already 4 ticks (5 in code but WaitTicks is weird). There's `DelayBeforeGameEnds` at 3 seconds, but it doesn't seem to do anything gameplay-wise since it comes after victory is already declared, so I'll reduce that by 2 seconds instead. I agree that the grace period should be minimized, although there is a discussion to be had considering the reward of killing a player and being able to abuse their idle units for 10 seconds (or however long). Which, if official, means there should be a patch for giving units shortly after dying before the victory condition detects defeat.

lL1l1 committed:
Fix victory condition not using DelayBeforeGameEnds

no functional change, it's already set to 3

lL1l1 committed:
Add time unit to EndGameGracePeriod comment

lL1l1 committed:
Shorten game end by 2 seconds

lL1l1 committed:
Merge branch 'develop' into simplify-defeat-backend

lL1l1 committed:
Remove duplicated functions

lL1l1 commented:
Another way to get all units to blow up is to have a player die before the game ends, and then have their last teammate die in the middle of the transfer. That could be another common scenario in games since one player leaving or dying can result in the rest of their team leaving shortly after. I think it's better to close up the PR though, this is "simplify defeat backend", not adding some mechanism that integrates army transfer on defeat and the overall match state.

lL1l1 committed:
Fix `EndGame` comment

lL1l1 committed:
Remove last GetArmyBrain

lL1l1 committed:
Use table.empty check instead of nil check for GiveUnitsToPlayer

lL1l1 committed:
Create fix.6985.md

Hdt80bro commented:
Yeah, that will have to wait for another time.
Everything looks good to me then!
